src/pylexibank/cldf.py

```diff
@@ -117,14 +117,15 @@ def add_form_with_segments(self, **kw):
         # Do we have morpheme segmentation on top of phonemes?
         with_morphemes = '+' in self['FormTable', 'Segments'].separator
 
-        language, concept, value, form, segments = [kw.get(v) for v in [
-            'Language_ID', 'Parameter_ID', 'Value', 'Form', 'Segments']]
+        language, concept, value, form, segments = (kw.get(v) for v in [
+            'Language_ID', 'Parameter_ID', 'Value', 'Form', 'Segments'])
 
         # check for required kws
-        if language is None or concept is None or value is None \
-                or form is None or segments is None:
+        if not all([language, concept, value, form, segments]):
             raise ValueError('language, concept, value, form, and segments must be supplied')
-        kw.setdefault('Segments', segments)
+
+        # Correct segments according to mapping in etc/segments.csv:
+        kw['Segments'] = [self.dataset.segments.get(s, s) for s in segments]
```

Comment:
We have been dealing with cases in which the number of segments in source and target replacements is not balanced, such as `"ah"` becoming `"a h"`. @LinguList solution was to join and split (i.e., `" ".join(segments).split()`) which is acceptable here.

My issue were potential cases where the number of segments would decrease, such as `"k w"` becoming `"kʷ"`, which is the reason why my code was more complex (joining *before* the replacement, performing the replacement by adding spaces, and *only then* stripping and splitting). This is far less elegant, and we didn't actually run into this kind of issue so far (probably one more case when I was planning too much ahead).

Reply:
Well, we could generalize this, and
- consider and read `SEGMENT` and `REPLACEMENT` as sub-sets of tokens,
- replace subsets with subsets.

Reply:
So how would that work? It would operate on the whitespaced segments?

Reply:
Btw.: Joining before replacing sub-lists doesn't seem acceptable to me, since it may **undo** proper segmentation, consider, e.g. `['ab', 'c']` and replacing `bc` with `x`.